libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,632 @@
+    }
+
+    pub fn from_hex(hex: &str) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber{ bn: BigInt::from_str_radix(hex, 16)? })
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber { bn: BigInt::from_bytes_be(Sign::Plus, bytes) })
+    }
+
+    pub fn to_dec(&self) -> Result<String, UrsaCryptoError> {
+        Ok(self.bn.to_str_radix(10))
+    }
+
+    pub fn to_hex(&self) -> Result<String, UrsaCryptoError> {
+        Ok(self.bn.to_str_radix(16).to_uppercase())
+    }
+
+    pub fn to_bytes(&self) -> Result<Vec<u8>, UrsaCryptoError> {
+        let (_, res) = self.bn.to_bytes_be();
+        Ok(res)
+    }
+
+    pub fn hash(data: &[u8]) -> Result<Vec<u8>, UrsaCryptoError> {
+        digest(DigestAlgorithm::Sha2_256, data).map_err(|e| UrsaCryptoError::InvalidStructure(e.to_string()))
+    }
+
+    pub fn add(&self, a: &BigNumber) -> Result<BigNumber, UrsaCryptoError> {
+        let res = &self.bn + &a.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn sub(&self, a: &BigNumber) -> Result<BigNumber, UrsaCryptoError> {
+        let res = &self.bn - &a.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn sqr(&self, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        let res = &self.bn * &self.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn mul(&self, a: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        let res = &self.bn * &a.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn mod_mul(&self, a: &BigNumber, n: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        //TODO: Use montgomery reduction
+        let res = (&self.bn * &a.bn) % &n.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn mod_sub(&self, a: &BigNumber, n: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        let res = (&self.bn - &a.bn) % &n.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn div(&self, a: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        let res = &self.bn / &a.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn add_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn += w;
+        Ok(self)
+    }
+
+    pub fn sub_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn -= w;
+        Ok(self)
+    }
+
+    pub fn mul_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn *= w;
+        Ok(self)
+    }
+
+    pub fn div_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn /= w;
+        Ok(self)
+    }
+
+    pub fn mod_exp(&self, a: &BigNumber, b: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        if a.is_negative() {
```

Comment:
`modpow` [panics]https://docs.rs/num-bigint/0.2.2/num_bigint/struct.BigInt.html#method.modpow() if modulus is less than 1 so a check for `b` is needed.